numpy/core/src/multiarray/common.h

```diff
@@ -292,6 +292,35 @@ npy_memchr(char * haystack, char needle,
     return p;
 }
 
+/*
+ * Helper to work around issues with the allocation strategy currently
+ * allocating not 1 byte for empty arrays, but enough for an array where
+ * all 0 dimensions are replaced with size 1 (if the itemsize is not 0).
+ *
+ * This means that we can fill in nice (nonzero) strides and still handle
+ * slicing direct math without being in danger of leaving the allocated byte
+ * bounds.
+ * In practice, that probably does not matter, but in principle this would be
+ * undefined behaviour in C.  Another solution may be to force the strides
+ * to 0 in these cases.  See also gh-15788.
+ *
+ * Unlike the code in `PyArray_NewFromDescr` does no overflow checks.
+ */
+static NPY_INLINE npy_intp
+PyArray_NBYTES_ALLOCATED(PyArrayObject *arr)
+{
+    if (PyArray_ITEMSIZE(arr) == 0) {
+        return 1;
+    }
+    npy_intp nbytes = PyArray_ITEMSIZE(arr);
```

Comment:
```suggestion
    npy_intp nbytes = PyArray_ITEMSIZE(arr);
    if (nbytes == 0) {
        return 1;
    }
```

Reply:
Do I need coffee, I don't see that it is different?  If `ITEMSIZE/`dtype->elsize` is 0, we always use 1 in both cases?  (Since that should be fine because strides are 0).  If itemsize/elsize is not 0, we need to multiply it with the shape (skipping 0 dims)?

Reply:
Ahh, I was just copying the code from the other places. So the difference is just to call PyArray_ITEMSIZE(arr) only once, not twice. I edited the suggestion, sorry for excessive copying.

Reply:
grr. still not quite right, editing again.